### PR TITLE
Plurals

### PR DESCRIPTION
**Pull request by aeng**
Plurals implementation:
https://bugzilla.redhat.com/show_bug.cgi?id=1172437

### CODE REVIEW

**Comment by lukebrooker** [comment]
:+1: Looks good to me but maybe @davidmason should look at the `slice()`s to be sure that's the best way to do it.

**Comment by davidmason** [comment]
I'll take a look this afternoon.

File: app/components/theme/lib/theme-map.css [hunk 1]
```diff
@@ -167,20 +168,20 @@
...
+  --Dropdown-content-color-bg    : var(--color-white);
+  --Dropdown-content-border      : var(--default-border);
+  --Dropdown-content-z-index     : var(--z-scale-8);
+  --Dropdown-title-size          : calc(14/16 * 1rem);
+  --Dropdown-title-color         : color(var(--color-secondary) tint(40%));
+  --Dropdown-title-color-bg      : transparent;
+  --Dropdown-item-color          : var(--color-highlight);
+  --Dropdown-item-hover-color    : var(--Dropdown-content-color-bg);
+  --Dropdown-item-hover-color-bg : var(--Dropdown-item-color);
+  --Dropdown-item-transition     : all .2s ease-out;
```
Comment (davidmason) [inline app/components/theme/lib/theme-map.css]: Is this being auto-formatted and rearranged? This could make for a lot of extra work in reviews if that is happening and we don't prevent it.

File: app/components/transUnit/TransUnitService.js [hunk 2]
```diff
@@ -22,10 +21,16 @@
...
     transUnitService.isTranslationModified = function(phrase) {
-      // on Firefox with input method turned on,
-      // when hitting tab it seems to turn undefined value into ''
-      return nullToEmpty(phrase.newTranslation) !== nullToEmpty(
-          phrase.translation);
+      if (phrase.plural) {
+        // on Firefox with input method turned on,
+        // when hitting tab it seems to turn undefined value into ''
+        return nullToEmpty(phrase.translations.toString()) !==
+          nullToEmpty(phrase.newTranslations.toString());
```
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: Converting the arrays to strings for comparison does not seem like good style to me (it is inefficient and may have edge cases where it fails). It is a bit shorter than other options, but lodash makes it pretty easy to compare each value:

```
var allSame = _.every(phrase.translations, function(translation, index) {
  return translation === phrase.newTranslations[index];
});
return !allSame;
```

Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: (wrap things in `nullToEmpty` as necessary)

File: app/components/transUnit/TransUnitService.js [hunk 3]
```diff
@@ -170,7 +198,26 @@
...
+      } else {
+        phrase.newTranslation = newText;
+      }
+      EventService.emitEvent(EventService.EVENT.TRANSLATION_TEXT_MODIFIED,
+        phrase);
+      EventService.emitEvent(EventService.EVENT.FOCUS_TRANSLATION,
+        phrase);
+    }
+
+    function modifyTranslationsText(phrase, newTexts) {
```
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: I'm sure we could simplify the code a lot by just treating everything the same whether it is plural or not. For anything that is not plural, just treat it as thought it is a plural with 1 source and 1 translation. There are lots of places in the code that we branch the behaviour, but there are only a few places we should need to treat them any different: receiving data from the server (populating the source/target arrays from either `content` or `contents`) and sending the data back to the server (putting the data in either `content` or `contents` have to be used correctly), and maybe a slightly different display template.

When can we simplify this code? I suspect there is insufficient time this sprint to make that much of a change.

File: app/components/transUnit/TransUnitService.js [hunk 4]
```diff
@@ -170,7 +198,26 @@
        updateSaveButton);
 
     function modifyTranslationText(phrase, newText) {
```
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: I think this would be better named something like `setTranslationText`, since it is not making any reference to the current value.

File: app/components/transUnit/TransUnitService.js (hunk 3 above)
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: If we treate everything as plurals, I think we would still weant this function, and we could name it `setAllTranslations` or similar.

File: app/components/theme/lib/theme-map.css (hunk 1 above)
Comment (lukebrooker) [inline app/components/theme/lib/theme-map.css]: Nope I did this manually. Well I use a shortcut for alignment but I rearranged manually. I'm slowly trying to make all my CSS this style. Most of it should be like this now.

File: app/components/transUnit/TransUnitService.js (hunk 3 above)
Comment (lukebrooker) [inline app/components/transUnit/TransUnitService.js]: I had the same sort of thought but as there were so many different things we are doing with translation, I was unsure of the best way to handle it. I am open to suggestions @davidmason

File: app/components/transUnit/TransUnitService.js (hunk 4 above)
Comment (lukebrooker) [inline app/components/transUnit/TransUnitService.js]: Sure, but this was not part of this PR. It existed already.

File: app/components/transUnit/TransUnitService.js (hunk 3 above)
Comment (aeng) [inline app/components/transUnit/TransUnitService.js]: I suggest we can remove content and source (singular) in phrase, and only have `contents` and `sources`. This should simplify lots of code as a start. About plural feature, not sure if it can make it in this sprint. Perhaps we can simplify all these in the plural branch.

File: app/components/transUnit/TransUnitService.js [hunk 5]
```diff
@@ -170,7 +198,26 @@
...
+      }
+      EventService.emitEvent(EventService.EVENT.TRANSLATION_TEXT_MODIFIED,
+        phrase);
+      EventService.emitEvent(EventService.EVENT.FOCUS_TRANSLATION,
+        phrase);
+    }
+
+    function modifyTranslationsText(phrase, newTexts) {
+      if(!phrase.plural) {
+        return; //only accept plural
```
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: This will silently fail if someone writes bad code that calls the wrong function. We should at least log an error so that the coder calling the wrong function can see why it is not working.

A more complete fix would be to make the function work regardless whether the phrase is a plural or not, which would be part of the suggestion in my earlier comment.

File: app/components/transUnit/TransUnitService.js [hunk 6]
```diff
@@ -170,7 +198,26 @@
...
+      }
+
+      //need slice(0) for new instance of array
+      phrase.newTranslations = newTexts.slice();
```
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: The comment is misleading, since the code does not use `slice(0)`. Can we change it to say `slice()`?

File: app/components/transUnit/TransUnitService.js [hunk 7]
```diff
@@ -190,7 +237,9 @@
     }
 
     function setSelected(transUnitCtrl, isSelected) {
-      transUnitCtrl.selected = isSelected || false;
+      if(transUnitCtrl.selected !== isSelected) {
+        transUnitCtrl.selected = isSelected || false;
+      }
```
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: Is there a problem with just setting the value to the same value? The extra check makes the code more complicated so I would like to know the reason for it, and we had better add a comment indicating what the reason is or someone might try to simplify it back to the previous structure later.

Comment (aeng) [inline app/components/transUnit/TransUnitService.js]: The checking is to prevent selected event being triggered repeatedly.

File: app/components/util/PhraseUtil.js [hunk 8]
```diff
@@ -6,16 +6,35 @@
...
+      if (phrase.plural) {
+        var newTranslations = phrase.newTranslations,
+            translations = phrase.translations;
+        if (_.isEmpty(_.compact(newTranslations))) {
+          return TransStatusService.getStatusInfo('untranslated');
+        }
+        else if (_.compact(newTranslations).length !== newTranslations.length) {
+          return TransStatusService.getStatusInfo('needswork');
+        }
+        else if (!_.isEmpty(_.difference(newTranslations, translations))) {
```
Comment (davidmason) [inline app/components/util/PhraseUtil.js]: This has a logical error: difference will be empty going from `[ "foo", "bar"]` to `[ "bar", "foo" ]`. There is another place that this comparison is being made (commented earlier with a lodash approach to it), so I think we need a function that will tell if 2 arrays of strings are the same. With a dedicated function like `haveTranslationsChanged(original, current)` we will ensure we don't hit this pitfall.

File: app/editor/EditorContentCtrl.js [hunk 9]
```diff
@@ -26,6 +23,42 @@
...
+          $location.search('status', null);
+        } else {
+          var queryString = '';
+
+          _.forEach(filter.status, function(val, key) {
+            if(val) {
+              queryString += key + ',';
+            }
+          });
+          queryString = queryString.substring(0, queryString.length - 1);
```
Comment (davidmason) [inline app/editor/EditorContentCtrl.js]: This looks like it is stripping the last character from `queryString`. Is that what is intended? If so, what is the purpose of doing so?

Comment (davidmason) [inline app/editor/EditorContentCtrl.js]: Oh wait, I just figured it out, it is getting rid of the `','`. I think we should instead put all the keys in an array (e.g. `keys`), then do `queryString = keys.join(',');`.

File: app/editor/EditorCtrl.js [hunk 10]
```diff
@@ -190,6 +220,45 @@
       EventService.emitEvent(EventService.EVENT.GOTO_PREV_PAGE);
     };
 
+    editorCtrl.resetFilter = function() {
+      resetFilter(true);
+    };
+
+    editorCtrl.updateFilter = function() {
+      updateFilter(true);
+    };
```
Comment (davidmason) [inline app/editor/EditorCtrl.js]: You could also encode the default value in the function definitions, so that here you could have `editorCrtl.updateFilter = updateFilter;`. You would just need to add a line at the start of the other function, like this:

```
function updateFilter(fireEvent) {
  fireEvent = arguments.length === 0 ? true : fireEvent;
```

Comment (davidmason) [inline app/editor/EditorCtrl.js]: Note: I don't necessarily think that approach is better, the current way seems a bit clearer.

File: app/editor/EditorShortcuts.js [hunk 11]
```diff
@@ -102,33 +83,33 @@
...
+    //   event.preventDefault();
+    //   event.stopPropagation();
+    //   if (editorShortcuts.selectedTUCtrl) {
+    //     EventService.emitEvent(EventService.EVENT.GOTO_NEXT_UNTRANSLATED,
+    //       currentContext());
+    //   }
+    //   // the shortcut is a tab + u combination
+    //   // we don't want other tab event to trigger
+    //   tabCombinationPressed = true;
+    // }
```
Comment (davidmason) [inline app/editor/EditorShortcuts.js]: Do we need to keep this commented out code?

File: app/editor/EditorShortcuts.js [hunk 12]
```diff
@@ -160,20 +141,15 @@
...
+        ['alt+k', 'alt+down']),
 
       GOTO_PREVIOUS_ROW: new ShortcutInfo(
         'mod+shift+enter', gotoPreviousRowCallback,
         'Save (if changed) and go to previous string',
-        ['alt+j', 'alt+up', 'shift+tab'])/*,
+        ['alt+j', 'alt+up'])/*,
         disable for now
       GOTO_NEXT_UNTRANSLATED: new ShortcutInfo(
         'tab+u', gotoToNextUntranslatedCallback)*/
```
Comment (davidmason) [inline app/editor/EditorShortcuts.js]: Disabled for how long? Why disabled?

File: app/editor/editor.html [hunk 13]
```diff
@@ -6,6 +6,33 @@
   <div ui-view="editor-suggestions"></div>
 
   <div ui-view="editor-details"></div>
+
+  <div class="Modal" ng-class="{'is-active': editor.showCheatsheet}" >
+    <div class="Modal-dialog">
+      <div class="Modal-header">
+        <h2 class="Modal-title">Keyboard Shortcuts</h2>
```
Comment (davidmason) [inline app/editor/editor.html]: I think this needs to be translatable.

File: app/components/theme/lib/theme-map.css (hunk 1 above)
Comment (davidmason) [inline app/components/theme/lib/theme-map.css]: Cool, that's fine then.

File: app/components/transUnit/TransUnitService.js (hunk 3 above)
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: I think the best way to do it is to convert the singular 'content' to an array as soon as we get it from the REST call, and convert it back just before it is sent to the server for a save. This would mean that most of the code can just assume there is an array and does not need to branch on plural/singular. In the few places that we need different display or logic, we can just use the `.isPlural` property to branch on.

File: app/components/transUnit/TransUnitService.js (hunk 4 above)
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: Still nice to fix :)

File: app/components/transUnit/TransUnitService.js (hunk 7 above)
Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: @aeng can you add a comment to say that, so that someone does not remove it?

Comment (davidmason) [inline app/components/transUnit/TransUnitService.js]: @aeng never mind, I was still looking at the outdated diff so I didn't see that you already added a comment.

**Comment by davidmason** [comment]
@aeng I think that commit fixes everything other than refactoring the singular/plural, so it should be ready to go to testing.

:+1:

**Comment by aeng** [comment]
Merged into master. https://github.com/zanata/zanata-spa/commit/b4449c9b5ddf0b5899c5f1f713d0bcf72c3c611e
